### ISSUE TITLE
Allow spilling to host memory in memory coloring

### ISSUE THREAD

**Issue by umangyadav** [ROCmSoftwarePlatform/AMDMIGraphX#1629]
Memory coloring currently assumes no limits on GPU memory and allocates as much space as necessary without checking for limit on it. 

Allow spilling to host memory if it goes beyond GPU  memory. 

This is useful for other backends like FPGA as well.

**Comment by umangyadav** [comment]
Spilling could be costly and could affect partitioning as well. 
Hide Spiling cycles by and restoring by other computations in parallel if possible.

**Comment by umangyadav** [comment]
Another issue is to allow users to provide device memory (MIGraphX doesn't have that ability). If so spilling needs to consider that and do it accordingly.

**Comment by umangyadav** [comment]
Re-open
